File: enode/src/main/java/com/qianzhui/enode/commanding/ICommandContext.java
package com.qianzhui.enode.commanding;

import com.qianzhui.enode.domain.IAggregateRoot;

import java.util.concurrent.CompletableFuture;

public interface ICommandContext {
    /**
     * Add a new aggregate into the current command context.
     *
     * @param aggregateRoot
     */
    void add(IAggregateRoot aggregateRoot);

    /**
     * Add a new aggregate into the current command context synchronously, and then return a completed task object.
     *
     * @param aggregateRoot
     * @return
     */
    CompletableFuture addAsync(IAggregateRoot aggregateRoot);

    /**
     * Get an aggregate from the current command context.
     *
     * @param <T>
     * @param id
     * @param firstFromCache
     * @return
     */
    <T extends IAggregateRoot> CompletableFuture getAsync(Object id, boolean firstFromCache, Class<T> clazz);

    void setResult(String result);

    String getResult();
}

File: enode/src/main/java/org/enodeframework/commanding/ICommandContext.kt
package org.enodeframework.commanding

import org.enodeframework.domain.IAggregateRoot
import org.enodeframework.messaging.IApplicationMessage
import java.util.concurrent.CompletableFuture

interface ICommandContext {
    /**
     * Add a new aggregate into the current command context.
     */
    fun add(aggregateRoot: IAggregateRoot?)

    /**
     * Add a new aggregate into the current command context synchronously, and then return a completed task object.
     */
    fun addAsync(aggregateRoot: IAggregateRoot?): CompletableFuture<Void?>?

    /**
     * Get an aggregate from the current command context.
     */
    fun <T : IAggregateRoot?> getAsync(id: Any?, firstFromCache: Boolean, clazz: Class<T>?): CompletableFuture<T>?
    fun <T : IAggregateRoot?> getAsync(id: Any?, clazz: Class<T>?): CompletableFuture<T>?
    var result: String?
    /**
     * Get an application message.
     */
    /**
     * Set an application message.
     */
    var applicationMessage: IApplicationMessage?
}